File: app/src/androidTest/java/id/djawadwipa/kalkulatorkeuangan/test/TestDocumentProvider.java
package id.djawadwipa.kalkulatorkeuangan.test;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.ParcelFileDescriptor;

import java.io.File;
import java.io.FileNotFoundException;

public final class TestDocumentProvider extends ContentProvider {
    @Override
    public boolean onCreate() {
        return true;
    }

    @Override
    public String getType(Uri uri) {
        return "application/octet-stream";
    }

    @Override
    public ParcelFileDescriptor openFile(Uri uri, String mode) throws FileNotFoundException {
        String lastPathSegment = uri.getLastPathSegment();
        if (lastPathSegment == null) {
            throw new FileNotFoundException("Nama dokumen tes tidak tersedia");
        }

        String sanitized = lastPathSegment.replaceAll("[^A-Za-z0-9._-]", "_");
        String fileName = sanitized.substring(0, Math.min(120, sanitized.length()));
        if (fileName.isEmpty()) {
            throw new FileNotFoundException("Nama dokumen tes tidak valid");
        }

        Context providerContext = getContext();
        if (providerContext == null) {
            throw new FileNotFoundException("Context provider tes tidak tersedia");
        }

        File root = new File(providerContext.getCacheDir(), "backup-smoke-test");
        if (!root.isDirectory() && !root.mkdirs()) {
            throw new FileNotFoundException("Direktori dokumen tes tidak dapat dibuat");
        }

        File file = new File(root, fileName);
        return ParcelFileDescriptor.open(file, ParcelFileDescriptor.parseMode(mode));
    }

    @Override
    public Cursor query(
            Uri uri,
            String[] projection,
            String selection,
            String[] selectionArgs,
            String sortOrder
    ) {
        return null;
    }

    @Override
    public Uri insert(Uri uri, ContentValues values) {
        return null;
    }

    @Override
    public int delete(Uri uri, String selection, String[] selectionArgs) {
        return 0;
    }

    @Override
    public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        return 0;
    }
}

File: app/src/androidTest/java/id/djawadwipa/kalkulatorkeuangan/test/TestDocumentProvider.kt
package id.djawadwipa.kalkulatorkeuangan.test

import android.content.ContentProvider
import android.content.ContentValues
import android.database.Cursor
import android.net.Uri
import android.os.ParcelFileDescriptor
import java.io.File

class TestDocumentProvider : ContentProvider() {
    override fun onCreate(): Boolean = true

    override fun getType(uri: Uri): String = "application/octet-stream"

    override fun openFile(uri: Uri, mode: String): ParcelFileDescriptor {
        val fileName = requireNotNull(uri.lastPathSegment)
            .replace(Regex("[^A-Za-z0-9._-]"), "_")
            .take(120)
        require(fileName.isNotBlank()) { "Nama dokumen tes tidak valid" }
        val root = File(requireNotNull(context).cacheDir, "backup-smoke-test").apply { mkdirs() }
        val file = File(root, fileName)
        return ParcelFileDescriptor.open(file, ParcelFileDescriptor.parseMode(mode))
    }

    override fun query(
        uri: Uri,
        projection: Array<out String>?,
        selection: String?,
        selectionArgs: Array<out String>?,
        sortOrder: String?,
    ): Cursor? = null

    override fun insert(uri: Uri, values: ContentValues?): Uri? = null

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<out String>?): Int = 0

    override fun update(
        uri: Uri,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<out String>?,
    ): Int = 0
}
